cpmpy/transformations/decompose_global.py

```diff
@@ -0,0 +1,100 @@
+import copy
+
+from ..expressions.globalconstraints import GlobalConstraint
+from ..expressions.core import Expression, Comparison, Operator
+from ..expressions.utils import is_any_list, eval_comparison
+from ..expressions.python_builtins import all
+from .flatten_model import flatten_constraint
+
+def decompose_global(lst_of_expr, supported=set(), supported_reif=set()):
+
+    """
+        Decomposes any global constraint not supported by the solver
+        Accepts a list of flat constraints as input
+        Returns a list of flat constraints
+
+        - supported: a set of supported global constraints or global functions
+        - supported_reified: a set of supported global constraints within a reification
+
+        Unsupported global constraints are decomposed into equivalent simpler constraints.
+        Numerical global constraints are left reified even if not part of the `supported_reified` set
+            as they can be rewritten using `cpmpy.transformations.reification.reify_rewrite`...
+                ... reified partial functions are decomposed when unsupported by `supported_reified`
+
+    """
+    def _is_supported(cpm_expr, reified):
+        if isinstance(cpm_expr, GlobalConstraint) and cpm_expr.is_bool():
+            if reified and cpm_expr.name not in supported_reif:
+                return False
+            if reified and cpm_expr.name not in supported:
+                return False
+        if isinstance(cpm_expr, Comparison) and isinstance(cpm_expr.args[0], GlobalConstraint):
+            if not cpm_expr.args[0].name in supported:
+                # reified numerical global constraints can be rewritten to non-reified versions
+                #  so only have to check for 'supported' set
+                return False
+            if reified and not cpm_expr.args[0].is_total() and cpm_expr.args[0].name not in supported_reif:
+                # edge case for partial functions as they cannot be rewritten to non-reified versions
+                #  have to decompose to total functions
+                #  ASSUMPTION: the decomposition for partial global functions is total! (for now only Element)
+                return False
+
+        return True
+
+    assert supported_reif <= supported, "`supported` set is assumed to be a subset of the `supported_reified` set"
```

Comment:
So something can be supported reified but not supported non-reified? That seems counter-intuitive.

Reply:
Other way around.. The test is correct but the message is not